### ISSUE TITLE
Unable to find remote state in 0.12 while it works on 0.11 using aws s3 backend

### ISSUE THREAD

**Issue by xeivieni** [hashicorp/terraform#22770]
### Terraform Version
```
v0.12.8
```

### Terraform Configuration Files
main.tf
```hcl
provider "aws" {
  assume_role {
    role_arn = "XXXXX"
  }
  region = "eu-west-1"
}

terraform {
  backend "s3" {
    bucket  = "my-bucket"
    key     = "myproject/main.tfstate"
    encrypt = "true"
    region         = "eu-west-1"
    role_arn = "XXXXX"
  }
}
```
datasources.tf
```hcl
data "terraform_remote_state" "remote_state" {
  backend = "s3"

  config = {
    bucket   = "my-bucket"
    region   = "eu-west-1"
    key      = "myproject/main.tfstate"
    role_arn = "XXXXX"
  }
}
```

The following commands succeed : `terraform init` `terraform workspace new staging`

At that time, my workspace is created and in my s3 bucket I can see my `main.tfstate` file at path : `my-bucket/env:/staging/myproject/`

### Crash Output
When I run `terraform plan` or `terraform apply`, I run into this issue : 
```
Error: Unable to find remote state

  on datasources.tf line 1, in data "terraform_remote_state" "remote_state":
   1: data "terraform_remote_state" "remote_state" {

No stored state was found for the given workspace in the given backend.
```

### Debug Output
```
-----------------------------------------------------
2019/09/11 18:12:06 [DEBUG] [aws-sdk-go] <?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">...</ListBucketResult>
2019/09/11 18:12:06 [DEBUG] [aws-sdk-go] DEBUG: Request s3/GetObject Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /my-bucket/myproject/main.tfstate HTTP/1.1
Host: s3.eu-west-1.amazonaws.com
User-Agent: aws-sdk-go/1.22.0 (go1.12.9; darwin; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.12.8
Authorization: AWS4-HMAC-SHA256 Credential=XXXXX/eu-west-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date;x-amz-security-token, Signature=XXXXX
X-Amz-Content-Sha256: XXXXXX
X-Amz-Date: 20190911T161206Z
X-Amz-Security-Token: XXXXXXX
Accept-Encoding: gzip


-----------------------------------------------------
2019/09/11 18:12:06 [DEBUG] [aws-sdk-go] DEBUG: Response s3/GetObject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 404 Not Found
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Wed, 11 Sep 2019 16:12:06 GMT
Server: AmazonS3
X-Amz-Id-2: XXXXXXX
X-Amz-Request-Id: XXXXXX


-----------------------------------------------------
2019/09/11 18:12:06 [DEBUG] [aws-sdk-go] <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Key>myproject/main.tfstate</Key><RequestId>XXXXX</RequestId><HostId>XXXXXX</HostId></Error>
2019/09/11 18:12:06 [DEBUG] [aws-sdk-go] DEBUG: Validate Response s3/GetObject failed, attempt 0/5, error NoSuchKey: The specified key does not exist.
	status code: 404, request id: XXXXXXX, host id: XXXXXXXXXXX
2019/09/11 18:12:06 [ERROR] <root>: eval: *terraform.EvalReadData, err: Unable to find remote state: No stored state was found for the given workspace in the given backend.
2019/09/11 18:12:06 [ERROR] <root>: eval: *terraform.EvalSequence, err: Unable to find remote state: No stored state was found for the given workspace in the given backend.
2019/09/11 18:12:06 [TRACE] [walkRefresh] Exiting eval tree: data.terraform_remote_state.remote_state
2019/09/11 18:12:06 [TRACE] vertex "data.terraform_remote_state.remote_state": visit complete
2019/09/11 18:12:06 [TRACE] vertex "data.terraform_remote_state.remote_state": dynamic subgraph encountered errors
2019/09/11 18:12:06 [TRACE] vertex "data.terraform_remote_state.remote_state": visit complete
2019/09/11 18:12:06 [TRACE] dag/walk: upstream of "provider.terraform (close)" errored, so skipping
2019/09/11 18:12:06 [TRACE] dag/walk: upstream of "root" errored, so skipping
```

### Expected Behavior
As the exact same configuration (appart from the `config = {` for terraform 0.12 compatibility) works under terraform 0.11, I expect it to work also for terraform 0.12 but it wont. 

### Steps to Reproduce
`terraform init`
`terraform apply`

### Additional Context
Under 0.11 I noticed that aws also returns the same 404 when trying to access this key in the bucket (as it does not exist) but this time, the output shows that terraform understands that the state has not been created yet and creates it. Output bellow :
``` 
2019/09/11 16:03:24 [DEBUG] [aws-sdk-go] <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Key>myproject/main.tfstate</Key><RequestId>XXXXXXX</RequestId><HostId>XXXXX</HostId></Error>
2019/09/11 16:03:24 [DEBUG] [aws-sdk-go] DEBUG: Validate Response s3/GetObject failed, not retrying, error NoSuchKey: The specified key does not exist.
	status code: 404, request id: XXXXX, host id: XXXXX
2019/09/11 16:03:24 [DEBUG] empty remote state
2019/09/11 16:03:24 [INFO] Resource 2019-09-11 14:03:24.928028 +0000 UTC has dynamic attributes
2019/09/11 16:03:24 [INFO] backend/local: apply calling Plan
2019/09/11 16:03:24 [INFO] terraform: building graph: GraphTypePlan
....
2019/09/11 16:03:24 [DEBUG] [aws-sdk-go] DEBUG: Request s3/PutObject Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /my-bucket/myproject/main.tfstate HTTP/1.1
```

### References
Haven't found any issue relating my problem.

**Comment by teamterraform** [comment]
Hi @xeivieni! Thanks for reporting this, and sorry for the surprising change in behavior.

This is unfortunately one of those situations where one person's bug is another person's feature. Terraform 0.12 intentionally checks that the state is present when retrieving state with `terraform_remote_state` because the usual interpretation of data sources is as assertions that something should already exist, and thus it's appropriate to return an error if the indicated object does not exist: that indicates that something has been applied out of order, and so usually better to return an error than to silently do something unexpected.

With that said then, the new behavior you've observed for Terraform 0.12 is intentional and not something we're planning to revert. The 0.11 behavior frequently caused confusion where people would make mistakes in their `config` objects and then see a strange failure downstream rather than a failure indicating the `data` block. The new behavior allows Terraform to give feedback in the correct location showing that it's the `config` object here that is wrong, not a downstream reference to the outputs of this data source.

The correct way to use `terraform_remote_state` (and most other data sources) is to apply the configuration that creates the state first and only then apply the one that consumes it. If we're understanding your example correctly here, it seems like you have the same configuration in both the `backend "s3"` block and in the `data "terraform_remote_state"` block, which is a strange thing to do: the data source here will always be one step behind the latest updates because the new state hasn't been updated yet at plan time. There is no reason to read the state for the configuration you're currently applying: all of the data you could obtain from it is already available directly from within the current configuration.

Because Terraform is working as designed here, we're going to close this now. There were a lot of different minor bug fixes in the Terraform 0.12.0 release that were impossible to enumerate completely, so we're sorry that this particular change did not appear as a specific line item in the release notes.

If you have any further questions, such as ideas for how to address your underlying problem in a different way that is 0.12-compatible, please feel free to start a new topic on [the community forum](https://discuss.hashicorp.com/c/terraform-core). Thanks!

**Comment by jonasv** [comment]
@teamterraform  What you did is a mistake here.

Because this is something what you have implemented very early and it used to work very well if done correctly and if you say that created a lot problem for the people who misused this feature then it is not tool's / terraform problem, it is people problem who use it in a different that the right way. So, now you have a situation where it affects I'm sure most of the community for the sake of the few, so we all could enjoy refactoring our stuff..

Great decision indeed.
